test/unit/org/apache/cassandra/net/MessagingServiceTest.java

```diff
@@ -53,4 +92,97 @@ public void testDroppedMessages()
         assertEquals(7500, (int)messagingService.getDroppedMessages().get(verb.toString()));
     }
+
+    @Test
+    public void shutdown_refusesNewMessagesWhenInProgress() throws InterruptedException {
+        Keyspace keyspace = Keyspace.open(KEYSPACE1);
+        DecoratedKey dk = Util.dk("key1");
+
+        Mutation mutation = new Mutation(KEYSPACE1, dk.getKey());
+        mutation.add(CF_STANDARD1, Util.cellname("Column1"), ByteBufferUtil.bytes("asdf"), 0);
+        MessageOut<Mutation> message = mutation.createMessage();
+        List<MessagingService.SocketThread> incomingAcceptThreads;
+        InetAddress oldBroadcast = FBUtilities.getBroadcastAddress();
+        try {
+            messagingService.listen();
+            assertTrue(messagingService.isListening());
+            assertFalse(MessagingService.instance().isListening());
+            incomingAcceptThreads = messagingService.getSocketThreads();
+            assertTrue(incomingAcceptThreads.size() > 0);
+
+            // Spoof broadcast address to avoid WriteCallbackInfo assertion
+            InetAddress mockBroadcast = mock(InetAddress.class);
+            doReturn(oldBroadcast.getAddress()).when(mockBroadcast).getAddress();
+            FBUtilities.setBroadcastInetAddress(mockBroadcast);
+
+            TestHandler handler = createHandler(keyspace);
+            MessagingService.instance().sendRR(message, FBUtilities.getLocalAddress(), handler, false);
+            handler.get();
+            assertEquals(1, handler.success);
+        } finally {
+            messagingService.shutdown();
+        }
+        incomingAcceptThreads.forEach(thread -> assertFalse(thread.isAlive()));
+        DatabaseDescriptor.setWriteRpcTimeout(Duration.ofSeconds(1).toMillis());
+        DatabaseDescriptor.setInternodeConnectionTimeout(Duration.ofMillis(50).toMillis());
+
+        Instant start = Instant.now();
+        TestHandler handler2 = createHandler(keyspace);
+        MessagingService.instance().sendRR(message, FBUtilities.getLocalAddress(), handler2, false);
+        handler2.get();
+        Duration handler2Time = Duration.between(start, Instant.now());
+        start = Instant.now();
+        TestHandler handler3 = createHandler(keyspace);
+        MessagingService.instance().sendRR(message, FBUtilities.getLocalAddress(), handler3, false);
+        handler3.get();
+        Duration handler3Time = Duration.between(start, Instant.now());
+        TestHandler handler4 = createHandler(keyspace);
+        MessagingService.instance().sendRR(message, FBUtilities.getLocalAddress(), handler4, false);
+        handler4.get();
+        Duration handler4Time = Duration.between(start, Instant.now());
+
+        // handler2 may or may not fail vs timeout. Likely due to OS level buffering, flushing the socket that has now
+        // been closed on the receiving end may or may not throw an IOException
```

Comment:
Could you give some more details about this? Based on your comment, handler2 could be expired by the callback reaper? Is the idea that the OS buffering would cause the `connect()` to still succeed?